src/api/procedures/utils.ts

```diff
@@ -322,3 +330,166 @@ export function assertRequirementsNotTooComplex(
+    case AuthorizationType.TransferTicker:
+      return assertTransferTickerAuthorizationValid(context, authRequest.data);
+    case AuthorizationType.TransferAssetOwnership:
+      return assertTransferAssetOwnershipAuthorizationValid(context, authRequest.data);
+    case AuthorizationType.BecomeAgent:
+      // no additional checks
+      return;
+    case AuthorizationType.AddMultiSigSigner:
+      // no additional checks
+      return;
+    case AuthorizationType.PortfolioCustody:
+      // no additional checks
+      return;
+    case AuthorizationType.JoinIdentity:
+      return assertJoinIdentityAuthorizationValid(authRequest);
+    case AuthorizationType.AddRelayerPayingKey:
+      return assertAddRelayerPayingKeyAuthorizationValid(authRequest.data);
+    default:
+      throw new UnreachableCaseError(authRequest.data); // ensures switch statement covers all values
+  }
+}
+
+/**
+ * asserts valid  primary key rotation authorization
+ */
+export async function assertPrimaryKeyRotationAuthorizationValid(
+  authRequest: AuthorizationRequest
+): Promise<void> {
+  if (authRequest.target instanceof Identity) {
+    throw new PolymeshError({
+      code: ErrorCode.ValidationError,
+      message: 'An Identity can not join another Identity',
+    });
+  }
+}
+
+/**
+ * asserts valid attest primary key authorization
+ */
+export async function assertAttestPrimaryKeyAuthorizationValid(
+  authRequest: AuthorizationRequest
+): Promise<void> {
+  if (!(await authRequest.issuer.isCddProvider())) {
+    throw new PolymeshError({
+      code: ErrorCode.UnmetPrerequisite,
+      message: 'Issuer must be a CDD provider',
+    });
+  }
+}
+
+/** asserts transfer ticker authorization is valid */
+export async function assertTransferTickerAuthorizationValid(
+  context: Context,
+  data: GenericAuthorizationData
+): Promise<void> {
+  const reservation = new TickerReservation({ ticker: data.value }, context);
+  const { status } = await reservation.details();
+  if (status === TickerReservationStatus.Free) {
+    throw new PolymeshError({
+      code: ErrorCode.UnmetPrerequisite,
+      message: 'The Ticker is not reserved',
+    });
+  }
+  if (status === TickerReservationStatus.TokenCreated) {
+    throw new PolymeshError({
+      code: ErrorCode.UnmetPrerequisite,
+      message: 'The ticker has already been used to create an Asset',
+    });
+  }
+}
+
+/** asserts valid transfer asset ownership authorization */
+export async function assertTransferAssetOwnershipAuthorizationValid(
+  context: Context,
+  data: GenericAuthorizationData
+): Promise<void> {
+  const token = new SecurityToken({ ticker: data.value }, context);
+  if (!(await token.exists()))
+    throw new PolymeshError({
+      code: ErrorCode.UnmetPrerequisite,
+      message: 'The Asset does not exist',
+    });
+}
+
+/**
+ * asserts valid join identity authorization request
+ */
+export async function assertJoinIdentityAuthorizationValid(
```

Comment:
This one should also check that the target is not an identity and is not already part of one

Reply:
Now that I think about it, a lot of authorizations are meant for either an identity or an account, but not both. We should probably have `assertTargetIsIdentity` and `assertTargetIsAccount`, and call them appropriately depending on the auth type

Reply:
Actually, as of version 3.7, you can do

```
function assertTargetIsIdentity(target: Signer): asserts target is Identity {
  // throw if not identity
}

// in function
assertTargetIsIdentity(target);

// if this code is reached, then the compiler knows `target` is an `Identity`
console.log(target.did);
```

More on assertion functions: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions

Reply:
I didn't know that was possible. I made an assertIsSigner function with that.